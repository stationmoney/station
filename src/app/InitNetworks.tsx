import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
