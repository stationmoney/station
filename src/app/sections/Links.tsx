import { useTranslation } from "react-i18next"
import DescriptionIcon from "@mui/icons-material/Description"
import BoltIcon from "@mui/icons-material/Bolt"
import { TUTORIAL, SETUP } from "config/constants"
import { ExternalLink } from "components/general"
import { Contacts } from "components/layout"
import styles from "./Links.module.scss"

const Links = () => {
  const { t } = useTranslation()

  const community = {
    medium: "https://medium.com/terra-money",
    discord: "https://terra.sc/discord",
    telegram: "https://t.me/TerraNetworkLobby",
    twitter: "https://twitter.com/terra_money",
    github: "https://github.com/terra-money",
  }

  return (
    <div className={styles.links}>
      <div className={styles.tutorial}>
        <ExternalLink href={SETUP} className={styles.link}>
          <BoltIcon style={{ fontSize: 18 }} />
          {t("Setup")}
        </ExternalLink>
        <ExternalLink href={TUTORIAL} className={styles.link}>
          <DescriptionIcon style={{ fontSize: 18 }} />
          {t("Documentation")}
        </ExternalLink>
      </div>

      <div className={styles.community}>
        <Contacts contacts={community} menu />
      </div>
    </div>
  )
}

export default Links
