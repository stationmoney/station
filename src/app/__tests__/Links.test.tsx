import React from "react"
import Links from "../sections/Links"
import { RecoilRoot } from "recoil"
import { NetworksProvider } from "../InitNetworks"
import { render } from "@testing-library/react"

function renderComponent() {
  type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
  const networks = {} as jest.Mocked<InterchainNetworks>
  const mockedTokenFilter = {} as jest.Mocked<TokenFilter>

  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <RecoilRoot>
        <Links />
      </RecoilRoot>
    </NetworksProvider>
  )
}

describe("Links component matches snapshots", () => {
  it("matches original component", () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
