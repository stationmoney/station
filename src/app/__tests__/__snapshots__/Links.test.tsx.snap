// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Links component matches snapshots matches original component 1`] = `
<DocumentFragment>
  <div
    class="links"
  >
    <div
      class="tutorial"
    >
      <a
        class="link"
        href="https://docs.terra.money/learn/station/download/station-desktop.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DescriptionIcon"
          focusable="false"
          style="font-size: 18px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
          />
        </svg>
        Documentation
      </a>
    </div>
    <div
      class="community"
    >
      <div
        class="flex start wrapper menu"
      >
        <a
          class="icon"
          href="https://medium.com/terra-money"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            height="18"
            width="18"
          >
            Medium.svg
          </svg>
        </a>
        <a
          class="icon"
          href="https://terra.sc/discord"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            height="18"
            width="18"
          >
            Discord.svg
          </svg>
        </a>
        <a
          class="icon"
          href="https://t.me/TerraNetworkLobby"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            height="18"
            width="18"
          >
            Telegram.svg
          </svg>
        </a>
        <a
          class="icon"
          href="https://twitter.com/terra_money"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            height="18"
            width="18"
          >
            Twitter.svg
          </svg>
        </a>
        <a
          class="icon"
          href="https://github.com/terra-money"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            height="18"
            width="18"
          >
            Github.svg
          </svg>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
